pkg/promotion/runner/builtin/set_freight_alias.go

```diff
@@ -0,0 +1,151 @@
+}
+
+// newSetFreightAlias returns an implementation of the promotion.StepRunner
+// interface that updates alias on Freight resources.
+func newSetFreightAlias(caps promotion.StepRunnerCapabilities) promotion.StepRunner {
+	return &setFreightAlias{
+		kargoClient:  caps.KargoClient,
+		schemaLoader: getConfigSchemaLoader(stepKindSetFreightAlias),
+	}
+}
+
+// Run implements the promotion.StepRunner interface.
+func (s *setFreightAlias) Run(
+	ctx context.Context,
+	stepCtx *promotion.StepContext,
+) (promotion.StepResult, error) {
+	cfg, err := s.convert(stepCtx.Config)
+	if err != nil {
+		return promotion.StepResult{
+			Status: kargoapi.PromotionStepStatusFailed,
+		}, &promotion.TerminalError{Err: err}
+	}
+
+	return s.run(ctx, stepCtx, cfg)
+}
+
+func (s *setFreightAlias) run(
+	ctx context.Context,
+	stepCtx *promotion.StepContext,
+	cfg builtin.SetFreightAliasConfig,
+) (promotion.StepResult, error) {
+	// This step intentionally targets a Freight specified explicitly via
+	// `freightID` instead of implicitly operating on
+	// `stepCtx.TargetFreightRef`.
+	//
+	// While most promotion steps act on the Freight currently being promoted,
+	// updating a Freight alias is a project-scoped mutation rather than a
+	// promotion-scoped one. There are valid use cases where a user may want to
+	// update the alias of a different Freight in the same project that is not
+	// currently being promoted
+	freight, err := api.GetFreightByNameOrAlias(ctx, s.kargoClient, stepCtx.Project, cfg.FreightID, "")
+	if err != nil {
+		return promotion.StepResult{
+				Status: kargoapi.PromotionStepStatusFailed,
+			}, &promotion.TerminalError{
+				Err: fmt.Errorf("failed to fetch Freight %q in project %q: %w", cfg.FreightID, stepCtx.Project, err),
+			}
+	}
+
+	if freight == nil {
+		return promotion.StepResult{
+				Status: kargoapi.PromotionStepStatusFailed,
+			}, &promotion.TerminalError{
+				Err: fmt.Errorf("freight %q not found in project %q", cfg.FreightID, stepCtx.Project),
+			}
+	}
+
+	// Check if alias is already used by another Freight in the project
+	freightList := kargoapi.FreightList{}
+	if err := s.kargoClient.List(
+		ctx,
+		&freightList,
+		client.InNamespace(stepCtx.Project),
+		client.MatchingLabels{kargoapi.LabelKeyAlias: cfg.NewAlias},
+	); err != nil {
+		return promotion.StepResult{
+				Status: kargoapi.PromotionStepStatusFailed,
+			}, &promotion.TerminalError{
+				Err: fmt.Errorf("failed to list Freights in project %q to check for alias conflicts: %w", stepCtx.Project, err),
+			}
+	}
+
+	if len(freightList.Items) > 1 || (len(freightList.Items) == 1 && freightList.Items[0].Name != freight.Name) {
+		return promotion.StepResult{
+				Status: kargoapi.PromotionStepStatusFailed,
+			}, &promotion.TerminalError{
+				Err: fmt.Errorf("alias %q is already in use by another Freight in project %q", cfg.NewAlias, stepCtx.Project),
+			}
+	}
```

Comment:
This shouldn't be required, as it would be captured by the validation webhook. Additionally, it's vulnerable to TOCTOU races: the alias could be claimed between the `List` and the `Patch`.

As the webhook enforces uniqueness, just let the `Patch` fail and surface that error.